lib/Runtime/Library/ArrayBuffer.cpp

```diff
@@ -834,6 +823,91 @@ namespace Js
     }
 #endif
 
+
+
+    WebAssemblyArrayBuffer::WebAssemblyArrayBuffer(uint32 length, DynamicType * type) :
+#ifndef ENABLE_FAST_ARRAYBUFFER
+        // Treat as a normal JavascriptArrayBuffer
+        JavascriptArrayBuffer(length, type) {}
+#else
+        JavascriptArrayBuffer(length, type, WasmVirtualAllocator)
+    {
+        // Make sure we always have a buffer even if the length is 0
+        if (buffer == nullptr)
+        {
+            // We want to allocate an empty buffer using virtual memory
+            Assert(length == 0);
+            buffer = (BYTE*)WasmVirtualAllocator(0);
+            if (buffer == nullptr)
+            {
+                JavascriptError::ThrowOutOfMemoryError(GetScriptContext());
+            }
+        }
+    }
+#endif
+
+    WebAssemblyArrayBuffer::WebAssemblyArrayBuffer(byte* buffer, uint32 length, DynamicType * type):
+        JavascriptArrayBuffer(buffer, length, type)
+    {
+
+    }
+
+    WebAssemblyArrayBuffer* WebAssemblyArrayBuffer::Create(byte* buffer, uint32 length, DynamicType * type)
+    {
+        Recycler* recycler = type->GetScriptContext()->GetRecycler();
+        WebAssemblyArrayBuffer* result;
+        if (buffer)
+        {
+            result = RecyclerNewFinalized(recycler, WebAssemblyArrayBuffer, buffer, length, type);
+        }
+        else
+        {
+            result = RecyclerNewFinalized(recycler, WebAssemblyArrayBuffer, length, type);
+        }
+        Assert(result);
+        recycler->AddExternalMemoryUsage(length);
+        return result;
+    }
+
+    bool WebAssemblyArrayBuffer::IsValidVirtualBufferLength(uint length)
+    {
+#if ENABLE_FAST_ARRAYBUFFER
+        return true;
+#else
+        return false;
+#endif
+    }
+
+    ArrayBuffer * WebAssemblyArrayBuffer::TransferInternal(uint32 newBufferLength)
+    {
+        if (newBufferLength == this->bufferLength)
+        {
+            return this;
```

Comment:
@MikeHolman I don't fully understand the spec for TransferInternal.
Must I always return a **new** ArrayBuffer or returning the current one is valid if the buffer doesn't change ?